plenum/test/consensus/view_change/test_view_change_service.py

```diff
@@ -69,34 +163,234 @@ def test_non_primary_responds_to_view_change_message_with_view_change_ack_to_new
-    service = view_change_service(name)
+    service = view_change_service_builder(name)
 
-    vc = view_change_message(initial_view_no + 1)
+    vc = create_view_change(initial_view_no)
     frm = some_item(validators, exclude=[name])
     service._network.process_incoming(vc, frm)
 
     assert len(service._network.sent_messages) == 0
 
 
-def test_new_view_message_is_sent_once_when_view_change_certificate_is_reached(
-        validators, primary, view_change_service, initial_view_no, view_change_message, view_change_acks):
+def test_new_view_message_is_sent_by_primary_when_view_change_certificate_is_reached(
+        internal_bus, validators, primary, view_change_service_builder, initial_view_no,
+        view_change_acks):
     primary_name = primary(initial_view_no + 1)
-    service = view_change_service(primary_name)
-    service.start_view_change()
+    service = view_change_service_builder(primary_name)
+
+    # start view change
+    internal_bus.send(NeedViewChange())
     service._network.sent_messages.clear()
 
+    # receive quorum of ViewChanges and ViewChangeAcks
     non_primaries = [item for item in validators if item != primary_name]
+    vc = create_view_change(initial_view_no)
     for vc_frm in non_primaries:
-        vc = view_change_message(initial_view_no + 1)
         service._network.process_incoming(vc, vc_frm)
-
         for ack, ack_frm in view_change_acks(vc, vc_frm, primary_name, len(validators) - 2):
             service._network.process_incoming(ack, ack_frm)
 
+    # check that NewView has been sent
     assert len(service._network.sent_messages) == 1
     msg, dst = service._network.sent_messages[0]
     assert dst is None  # message was broadcast
     assert isinstance(msg, NewView)
     assert msg.viewNo == initial_view_no + 1
+
+
+def test_new_view_message_is_not_sent_by_non_primary_when_view_change_certificate_is_reached(
+        internal_bus, validators, primary, view_change_service_builder, initial_view_no, some_item):
+    next_view_no = initial_view_no + 1
+    primary_name = primary(next_view_no)
+    non_primary_name = some_item(validators, exclude=[primary_name])
+    service = view_change_service_builder(non_primary_name)
+
+    # start view change
+    internal_bus.send(NeedViewChange())
+    service._network.sent_messages.clear()
+
+    # receive quorum of ViewChanges and ViewChangeAcks
+    non_primaries = [item for item in validators if item != primary_name]
+    vc = create_view_change(initial_view_no)
+    for vc_frm in non_primaries:
+        service._network.process_incoming(vc, vc_frm)
+
+        for ack, ack_frm in create_view_change_acks(vc, vc_frm, non_primaries):
+            service._network.process_incoming(ack, ack_frm)
+
+    # check that NewView hasn't been sent
+    assert all(not isinstance(msg, NewView) for msg in service._network.sent_messages)
+
+
+def test_view_change_finished_is_sent_by_primary_once_view_change_certificate_is_reached(internal_bus, validators,
```

Comment:
Just small note - theoretically acquiring view change certificate doesn't guarantee that view change can be completed. If there were some view changes from malicious nodes then primary might not be able to build a new view message on just n-f message (but it will be able to build it after receiving n-f messages from honest nodes).